src/KurrentDB.SecondaryIndexing/Indexes/Category/CategoryIndexProcessor.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Kurrent, Inc and/or licensed to Kurrent, Inc under one or more agreements.
+// Kurrent, Inc licenses this file to you under the Kurrent License v1 (see LICENSE.md).
+
+using KurrentDB.Core.Data;
+using KurrentDB.SecondaryIndexing.Storage;
+using static KurrentDB.SecondaryIndexing.Indexes.Category.CategorySql;
+
+namespace KurrentDB.SecondaryIndexing.Indexes.Category;
+
+internal class CategoryIndexProcessor {
+	readonly Dictionary<string, int> _categories;
+	readonly Dictionary<int, long> _categorySizes = new();
+	readonly DuckDbDataSource _db;
+
+	int _seq;
+	public long LastIndexesPosition { get; private set; }
+
+	public CategoryIndexProcessor(DuckDbDataSource db) {
+		_db = db;
+		using var connection = db.OpenNewConnection();
+		var ids = connection.Query<ReferenceRecord, GetCategoriesQuery>();
+		_categories = ids.ToDictionary(x => x.Name, x => x.Id);
+
+		foreach (var id in ids) {
+			_categorySizes[id.Id] = -1;
+		}
+
+		var sequences = connection.Query<(int Id, long Sequence), GetCategoriesMaxSequencesQuery>();
+		foreach (var sequence in sequences) {
+			_categorySizes[sequence.Id] = sequence.Sequence;
+		}
+
+		_seq = _categories.Count > 0 ? _categories.Values.Max() : 0;
+	}
+
+	public SequenceRecord Index(ResolvedEvent resolvedEvent) {
+		var categoryName = GetStreamCategory(resolvedEvent.OriginalStreamId);
+
+		if (_categories.TryGetValue(categoryName, out var categoryId)) {
+			var next = _categorySizes[categoryId] + 1;
+			_categorySizes[categoryId] = next;
+			LastIndexesPosition = resolvedEvent.Event.LogPosition;
+			return new(categoryId, next);
+		}
+
+		var id = ++_seq;
+
+		_categories[categoryName] = id;
+		_categorySizes[id] = 0;
+
+		_db.Pool.ExecuteNonQuery<AddCategoryStatementArgs, AddCategoryStatement>(new(id, categoryName));
+		LastIndexesPosition = resolvedEvent.Event.LogPosition;
+		return new(id, 0);
+	}
+
+	public long GetLastEventNumber(int categoryId) =>
+		_categorySizes.TryGetValue(categoryId, out var size) ? size : ExpectedVersion.NoStream;
+
+	public int GetCategoryId(string categoryName) =>
+		_categories.TryGetValue(categoryName, out var categoryId) ? categoryId : -1;
+
+	private static string GetStreamCategory(string streamName) {
```

Comment:
suggestion: rename to `TryGetStreamCategory(string streamName, out string category)` for better semantics